Admin Defined Node Disruptions: Tech Preview (#74562)
openshift/openshift-docs: modules/machine-config-node-disruption-config.adoc

// Module included in the following assemblies:
//
// * post_installation_configuration/machine-configuration-tasks.adoc

:_mod-docs-content-type: PROCEDURE
[id="machine-config-node-disruption-config_{context}"]
= Configuring node restart behaviors upon machine config changes

You can create a node disruption policy to define the machine configuration changes that cause a disruption to your cluster, and which changes do not.

You can control how your nodes respond to changes in the files in the `/var` or `/etc` directory, the systemd units, the SSH keys, and the `registries.conf` file.

include::snippets/machine-config-node-disruption-actions.adoc[]

.Prerequisite

> 

Reviewer: remove extra blank line

> * You have enabled the `TechPreviewNoUpgrade` feature set by using the feature gates. For more information, see "Enabling features using feature gates" in the Additional resources of this section.

Reviewer: I think they recently revised the guidance and said not to use the wording "in the additional resources section" - https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#lead-in-link-sentence-style-for-non-xref-situations

Author: @kcarmichael08 That's what I thought. I only added that after someone in a different PR told me to add it -- yesterday!

====

.Procedure

. Edit the `machineconfigurations.operator.openshift.io` object to define the node disruption policy:
+
[source,terminal]
----
$ oc edit MachineConfiguration cluster -n openshift-machine-config-operator
----

. Add a node disruption policy similar to the following:
+
[source,yaml]
----
apiVersion: operator.openshift.io/v1
kind: MachineConfiguration
metadata:
  name: cluster
# ...
spec:
  nodeDisruptionPolicy: <1>
    files: # <2>
    - actions: # <3>
      - reload: # <4>
          serviceName: chronyd.service # <5>
        type: Reload
      path: /etc/chrony.conf # <6>
    sshkey: # <7>
      actions:
      - type: Drain
      - reload:
          serviceName: crio.service
        type: Reload
      - type: DaemonReload
      - restart:
          serviceName: crio.service
        type: Restart
    units: # <8>
    - actions:
      - type: Drain
      - reload:
          serviceName: crio.service
        type: Reload
      - type: DaemonReload
      - restart:
          serviceName: crio.service
        type: Restart
      name: test.service
----
<1> Specifies the node disruption policy. 

> <2> Specifies a list of machine config file definitions and actions to take to changes on those paths This list supports a maximum of 50 entries.

Reviewer: ```suggestion
<2> Specifies a list of machine config file definitions and actions to take to changes on those paths. This list supports a maximum of 50 entries.
```

<3> Specifies the series of actions to be executed upon changes to the specified files. Actions are applied in the order that they are set in this list. This list supports a maximum of 10 entries.
<4> Specifies that the listed service is to be reloaded upon changes to the specified files. 
<5> Specifies the full name of the service to be acted upon. 
<6> Specifies the location of a file that is managed by a machine config. The actions in the policy apply when changes are made to the file in `path`.
<7> Specifies a list of service names and actions to take upon changes to the SSH keys in the cluster. This list supports a maximum of 50 entries.
<8> Specifies a list of systemd unit names and actions to take upon changes to those units. This list supports a maximum of 50 entries.

.Verification

> . View the `MachineConfiguration` object file that you created:

Reviewer: ```suggestion
* View the `MachineConfiguration` object file that you created:
```
per https://redhat-documentation.github.io/supplementary-style-guide/#single-step-procedures

// Module included in the following assemblies:
//
// * post_installation_configuration/machine-configuration-tasks.adoc

:_mod-docs-content-type: PROCEDURE
[id="machine-config-node-disruption-config_{context}"]
= Configuring node restart behaviors after machine config changes

You can create a node disruption policy to define the machine configuration changes that cause a disruption to your cluster, and which changes do not.

You can control how your nodes respond to changes in the configuration files in the `/var` or `/etc` directory, systemd units, or SSH keys.

include::snippets/machine-config-node-disruption-actions.adoc[]

.Prerequisite

* You have enabled the `TechPreviewNoUpgrade` feature set by using the feature gates. For more information, see _Enabling features using feature gates_.
+
[WARNING]
====

> Enabling the `TechPreviewNoUpgrade` feature set on your cluster prevents minor version updates. The `TechPreviewNoUpgrade` feature set cannot be disabled. Do not enable this feature set on production clusters.

Reviewer 2: @mburke5678 what does the minor version mean here, y or z? as I know z-stream level upgrade is allowed.

Author: @rioliu-rh I don't know, to be honest. This is standard boilerplate wording that we use. I believe it means y-stream updates.

Reviewer 2: you're right, the version pattern is like major.minor.patch
xref: https://www.redhat.com/en/blog/the-ultimate-guide-to-openshift-update-for-cluster-administrators

====

.Procedure

> . Edit the `MachineConfiguration` object to define the node disruption policy:

Reviewer 2: suggest to use full name like `machineconfigurations.operator.openshift.io`

+
[source,terminal]
----
$ oc edit MachineConfiguration cluster -n openshift-machine-config-operator
----

. Add a node disruption policy similar to the following:
+
[source,yaml]
----
apiVersion: operator.openshift.io/v1
kind: MachineConfiguration
metadata:
  name: cluster
  namespace: openshift-machine-config-operator
# ...
spec:
  nodeDisruptionPolicy: <1>
    files: # <2>
    - actions: # <3>
      - reload: # <4>
          serviceName: chronyd.service # <5>
        type: Reload
      path: /etc/chrony.conf # <6>
    sshkey: # <7>
      actions:
      - type: Drain
      - reload:
          serviceName: crio.service
        type: Reload
      - type: DaemonReload
      - restart:
          serviceName: crio.service
        type: Restart
    units: # <8>
    - actions:
      - type: Drain
      - reload:
          serviceName: crio.service
        type: Reload
      - type: DaemonReload
      - restart:
          serviceName: crio.service
        type: Restart
      name: test.service
----

> <1> Specifies the node disruption policy for the nodes associated with the role label you configured. 

Reviewer 2: NDP is global resource, it is not associated with any node or mcp.

You can create a node disruption policy to define the machine configuration changes that cause a disruption to your cluster, and which changes do not.

You can control how your nodes respond to changes in the configuration files in the `/var` or `/etc` directory, systemd units, or SSH keys.

include::snippets/machine-config-node-disruption-actions.adoc[]

.Prerequisite

* You have enabled the `TechPreviewNoUpgrade` feature set by using the feature gates. For more information, see _Enabling features using feature gates_.
+
[WARNING]
====
Enabling the `TechPreviewNoUpgrade` feature set on your cluster prevents minor version updates. The `TechPreviewNoUpgrade` feature set cannot be disabled. Do not enable this feature set on production clusters.
====

.Procedure

. Edit the `MachineConfiguration` object to define the node disruption policy:
+
[source,terminal]
----
$ oc edit MachineConfiguration cluster -n openshift-machine-config-operator
----

. Add a node disruption policy similar to the following:
+
[source,yaml]
----
apiVersion: operator.openshift.io/v1
kind: MachineConfiguration
metadata:
  name: cluster
  namespace: openshift-machine-config-operator
  labels:

>     machineconfiguration.openshift.io/role: worker # <1>

Reviewer 3: This configuration is not role specific, the policy is applied cluster wide, to all nodes. So this label would not do anything.

# ...
spec:
  nodeDisruptionPolicy: <2>
    files: # <3>
    - actions: # <4>
      - reload: # <5>
          serviceName: chronyd.service # <6>
        type: Reload
      path: /etc/chrony.conf # <7>
    sshkey: # <8>
      actions:
      - type: Drain
      - reload:
          serviceName: crio.service
        type: Reload
      - type: DaemonReload
      - restart:
          serviceName: crio.service
        type: Restart
    units: # <9>
    - actions:
      - type: Drain
      - reload:
          serviceName: crio.service
        type: Reload
      - type: DaemonReload
      - restart:
          serviceName: crio.service
        type: Restart
      name: test.service
----

> <1> Specifies the role label for the custom machine config pool where you want to use the policy.

Reviewer 3: same comment as above

<2> Specifies the node disruption policy for the nodes associated with the role label you configured. 
<3> Specifies a list of machine config file definitions and actions to take to changes on those paths This list supports a maximum of 50 entries.
<4> Specifies the series of action to be executed upon changes to the specified files. Actions are applied in the order that they are set in this list. This list supports a maximum of 10 entries.
<5> Specifies that the specified service is to be reloaded upon changes to the specified files. 
<6> Specifies the full name of the service to be acted upon. 
<7> Specifies the location of a file that is managed by a machine config. The actions in the policy apply when changes are made to the file in `path`.
<8> Specifies an action all changes to the SSH key changes in the cluster.
<9> Specifies a list of systemd unit names and actions to take upon changes to those units. This list supports a maximum of 50 entries.

.Verification

> . View the `MachineConfig` object file that you created:

Reviewer 3: ```suggestion
. View the `MachineConfiguration` object file that you updated:
```
A default MachineConfiguration object will exist before any changes are applied, showing the standard MCO policies. It may be good to point those out here as well. More info here: https://github.com/openshift/machine-config-operator/blob/master/docs/NodeDisruptionPolicy.md#how-it-works-right-now
A user will always be updating it with their custom policies(overriding these standard policies, if there is a conflict)

+
----

> $ oc edit mc node-disruption-policy

Reviewer 3: ```suggestion
$ oc get MachineConfiguration/cluster -o yaml
```
